source/Halibut.Tests/BiDirectionalTests.cs

```diff
@@ -0,0 +1,182 @@
+            
+            // Since EndpointB is not polling, it will not receive the message after waiting some time
+            Func<Task> taskA = () => clientServiceA.SayHelloAsync("Hello");
+            await taskA.Should().ThrowAsync<HalibutClientException>();
+            
+            // Now start polling from EndpointB
+            var listeningEndpoint = new ServiceEndPoint("https://localhost:" + port, _endpointAThumbprint, timeoutsAndLimits);
+            halibutRuntimeB.Poll(polling, listeningEndpoint, CancellationToken);
+            
+            // ClientA can now send the message
+            var result2 = await clientServiceA.SayHelloAsync("World").WithCancellation(CancellationToken);
+            result2.Should().Be("World...");
+
+            // Disconnect EndpointB, ClientA should no longer send message
+            await halibutRuntimeB.DisconnectAsync(listeningEndpoint, CancellationToken);
+            Func<Task> taskA2 = () => clientServiceA.SayHelloAsync("Hello");
+            await taskA2.Should().ThrowAsync<HalibutClientException>();
+            
+            // EndpointB Reconnects so EndpointA can now be called
+            halibutRuntimeB.Poll(polling, listeningEndpoint, CancellationToken);
+            result2 = await clientServiceA.SayHelloAsync("World").WithCancellation(CancellationToken);
+            result2.Should().Be("World...");
+        }
+
+        
+        HalibutRuntime CreateSimpleHalibutRuntime(X509Certificate2 certificate2)
+        {
+            var services = new DelegateServiceFactory();
+            services.Register<IMyEchoService, IAsyncMyEchoService>(() => new MyEchoService());
+            return new HalibutRuntimeBuilder()
+                .WithServerCertificate(certificate2)
+                .WithServiceFactory(services)
+                .WithHalibutTimeoutsAndLimits(new HalibutTimeoutsAndLimitsForTestsBuilder().Build())
+                .Build();
+        }
+
+        public interface IMyEchoService
+        {
+            void BlockWaitingForSharedStateIncrement();
+            string SayHello(string name);
+        }
+
+        public interface IAsyncClientMyEchoService
+        {
+            Task BlockWaitingForSharedStateIncrementAsync();
+            Task<string> SayHelloAsync(string name);
+        }
+
+        public interface IAsyncMyEchoService
+        {
+            Task BlockWaitingForSharedStateIncrementAsync(CancellationToken cancellationToken);
+            Task<string> SayHelloAsync(string name, CancellationToken cancellationToken);
+        }
+
+        public class MyEchoService : IAsyncMyEchoService
+        {
+            public static int SharedState = 0;
+            public async Task BlockWaitingForSharedStateIncrementAsync(CancellationToken cancellationToken)
+            {
+                SharedState++;
```

Comment:
nit: I always wondered if ints need to be interlocked incremented to be safe, so I end up doing to be safe but maybe we don't need to?

Reply:
Yeah i think this is probably a good case.
Since this effectively amounts to
`SharedState = SharedState + 1` there is room for threading issues. :)